gpkit/geometric_program.py

```diff
@@ -382,235 +101,182 @@ def _solve(self, solver=None, printing=True, skipfailures=False,
             solverfn = solver
             solver = solver.__name__
         else:
-            if not solver:
-                raise ValueError("No solver was given; perhaps gpkit was not"
-                                 " properly installed, or found no solvers"
-                                 " during the install process.")
-            raise ValueError("Solver %s is not implemented!" % solver)
-        self.solverfn = solverfn
-        self.solver = solver
+            raise ValueError("Unknown solver '%s'." % solver)
 
-        if printing:
+        if verbosity > 0:
             print("Using solver '%s'" % solver)
-        self.starttime = time()
+            self.starttime = time()
+            print("Solving for %i variables." % len(self.varlocs))
+
+        solver_out = solverfn(self.cs, self.A, self.p_idxs, self.k,
+                              *args, **kwargs)
+
+        self.solver_out = solver_out  # NOTE: SIDE EFFECTS AHOY
+        # TODO: add a solver_log file using stream debugger
+
+        if verbosity > 0:
+            print("Solving took %.3g seconds." % (time() - self.starttime))
 
-        if self.sweep:
-            solution = self._solve_sweep(printing,
-                                         skipfailures,
-                                         allownonoptimal)
+        result = {}
+        result["variables"] = dict(zip(self.varlocs,
+                                   np.exp(solver_out['primal']).ravel()))
```

Comment:
I'm confused about something very basic. `self.varlocs` is a dict. I assume `solver_out['primal']` is a list or tuple. How on earth do we know that the order of the dict keys matches the order in `solver_out['primal']`?